File: Sources/TidalEngine/Station.swift
import Foundation

/// A tide station: location plus the harmonic constituents needed to
/// predict water level there.
public struct Station: Equatable, Sendable {
    /// Stable identifier from the source dataset (e.g. NOAA/CHS station ID).
    public let id: String

    public let name: String
    public let latitude: Double
    public let longitude: Double

    /// Height (in the same units as each constituent's amplitude) that
    /// predictions are measured relative to, e.g. Mean Lower Low Water.
    public let datumOffset: Double

    public let constituents: [Constituent]

    public init(
        id: String,
        name: String,
        latitude: Double,
        longitude: Double,
        datumOffset: Double,
        constituents: [Constituent]
    ) {
        self.id = id
        self.name = name
        self.latitude = latitude
        self.longitude = longitude
        self.datumOffset = datumOffset
        self.constituents = constituents
    }
}

File: engine/src/station.rs
use crate::Constituent;

/// A tide station: location plus the harmonic constituents needed to
/// predict water level there.
#[derive(Debug, Clone, PartialEq)]
pub struct Station {
    /// Stable identifier from the source dataset (e.g. NOAA/CHS station ID).
    pub id: String,
    pub name: String,
    pub latitude: f64,
    pub longitude: f64,
    /// Height (same units as each constituent's amplitude) that
    /// predictions are measured relative to, e.g. Mean Lower Low Water.
    pub datum_offset: f64,
    pub constituents: Vec<Constituent>,
}

impl Station {
    pub fn new(
        id: impl Into<String>,
        name: impl Into<String>,
        latitude: f64,
        longitude: f64,
        datum_offset: f64,
        constituents: Vec<Constituent>,
    ) -> Self {
        Self {
            id: id.into(),
            name: name.into(),
            latitude,
            longitude,
            datum_offset,
            constituents,
        }
    }
}
